codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -317,85 +315,98 @@ private Http2Stream requireStream(int streamId) {
-                        writePromise = promise;
+            int maxFrameSize = frameWriter().configuration().frameSizePolicy().maxFrameSize();
+            do {
+                int allowedFrameSize = Math.min(maxFrameSize, allowedBytes - bytesWritten);
+
+                // Always need a real promise for write as we have to listen to it for failure.
+                ChannelPromise promise = ctx.newPromise();
+                promise.addListener(this);
+                CompositeByteBuf toWrite = ctx.alloc().compositeBuffer(bufsAndPromises.size());
+                while (!bufsAndPromises.isEmpty()) {
+                    final Object entry = bufsAndPromises.peek();
+                    if (entry instanceof ByteBuf) {
+                        ByteBuf buf = (ByteBuf) entry;
+                        if (toWrite.readableBytes() + buf.readableBytes() > allowedFrameSize) {
+                            ByteBuf slice = buf.readSlice(allowedFrameSize - toWrite.readableBytes());
+                            toWrite.addComponent(slice);
+                            toWrite.writerIndex(toWrite.writerIndex() + slice.readableBytes());
+                            // The remaining buffer is kept as a retained slice of the original
+                            bufsAndPromises.remove();
+                            bufsAndPromises.addFirst(buf.slice().retain());
+                            break;
+                        } else {
+                            toWrite.addComponent(buf);
+                            toWrite.writerIndex(toWrite.writerIndex() + buf.readableBytes());
+                            bufsAndPromises.remove();
+                        }
                     } else {
-                        // Create a new promise and listen to it for failure
-                        writePromise = ctx.newPromise();
-                        writePromise.addListener(this);
-                    }
-                    if (toWrite instanceof SlicedByteBuf && data instanceof CompositeByteBuf) {
-                        // If we're writing a subset of a composite buffer then we want to release
-                        // any underlying buffers that have been consumed. CompositeByteBuf only releases
-                        // underlying buffers on write if all of its data has been consumed and its refCnt becomes
-                        // 0.
-                        final CompositeByteBuf toFree = (CompositeByteBuf) data;
-                        writePromise.addListener(new ChannelFutureListener() {
-                            @Override
-                            public void operationComplete(ChannelFuture future) throws Exception {
-                                toFree.discardReadComponents();
-                            }
-                        });
+                        bufsAndPromises.remove();
+                        final ChannelPromise nested = (ChannelPromise) entry;
+                        if (!nested.isVoid()) {
```

Comment:
Seems like you are preventing void promises from being added to `bufsAndPromises`. Is this conditional check necessary?